internal/handlers/pg/pgdb/query.go

```diff
@@ -268,61 +267,133 @@ func prepareWhereClause(sqlFilters *types.Document) (string, []any, error) {
+				}
 			}
 
-			continue
-		}
-
-		switch v := v.(type) {
-		case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+		case *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
 			// type not supported for pushdown
-			// TODO $eq and $ne https://github.com/FerretDB/FerretDB/issues/1840
-			// TODO $gt and $lt https://github.com/FerretDB/FerretDB/issues/1875
 			continue
 
 		case float64, string, types.ObjectID, int32, int64:
-			// Select if value under the key is equal to provided value.
-			// If the value under the key is not equal to v,
-			// but the value under the key k is an array - select if it contains the value equal to v.
-			sql := `(_jsonb->%[1]s)::jsonb = %[2]s OR (_jsonb->%[1]s)::jsonb @> %[2]s`
-
-			// placeholder p.Next() returns SQL argument references such as $1, $2 to prevent SQL injections.
-			// placeholder $1 is used for field key,
-			// placeholder $2 is used for field value v.
-			filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next()))
-			args = append(args, k, string(must.NotFail(pjson.MarshalSingleValue(v))))
+			builder.addFilter(k, eq, v)
 
 		default:
 			panic(fmt.Sprintf("Unexpected type of value: %v", v))
 		}
 	}
 
-	var query string
+	query, args := builder.generateWhereClause()
+
+	return query, args, nil
+}
+
+// builderOperator represents available filtersBuilder operators.
+type builderOperator uint8
+
+const (
+	eq builderOperator = iota
+	ne
+)
 
-	if len(filters) > 0 {
-		query = ` WHERE ` + strings.Join(filters, " AND ")
+// filtersBuilder is responsible for building SQL filters.
+// It allows to procedurally add simple key-value comparisons,
+// and translate them to the single SQL WHERE clause.
+type filtersBuilder struct {
+	filters []string
+	args    []any
+	p       Placeholder
+}
+
+// addFilter creates SQL filter based on provided key, value and operator.
+func (fb *filtersBuilder) addFilter(key string, op builderOperator, val any) {
+	// check if values are equal or if the left value contains the right one
+	eqOperator := "@>"
+
+	if key == "_id" {
+		// check if values are equal
+		eqOperator = "="
 	}
 
-	return query, args, nil
+	switch op {
+	case eq:
+		// Select if value under the key is equal to provided value.
+		sql := `(_jsonb->%[1]s)::jsonb ` + eqOperator + ` %[2]s`
+
+		fb.filters = append(fb.filters, fmt.Sprintf(sql, fb.p.Next(), fb.p.Next()))
+		fb.args = append(fb.args, key, string(must.NotFail(pjson.MarshalSingleValue(val))))
+
+	case ne:
+		sql := `NOT ( ` +
+			// does document contain the key,
+			// it is necessary, as NOT won't work correctly if the key does not exist.
+			`_jsonb ? %[1]s AND ` +
+
+			// does the value under the key is equal to the filter
+			`(_jsonb->%[1]s)::jsonb ` + eqOperator + ` %[2]s AND ` +
+
+			// does the value type is equal to the filter's one
+			`(_jsonb->'$s'->'p'->%[1]s->'t')::jsonb = '"` + pjson.GetTypeOfValue(val) +
```

Comment:
So I'm not sure if this is working. For example, if val is `int(1)`, db has `double(1)` and `long(1)` they are considered equal. For example, `{v: {$ne: 1}` on values `{v: int(1)}`, `{v: double(1)}` and `{v: long(1)}` should be false. 
But tests in integration are passing so I don't think I'm understanding how it works.

Reply:
Thanks for spotting that! I'm investigating that

Reply:
I've inserted those documents:
```
db.values.insertMany([{v:"1"},{v:1},{v:Int32("1")},{v:1.0},{v:Long("1")},{v:Double("1")}])
```
and it seems that they are converted to the decimal.
```
test> db.values.find()
[
  { _id: ObjectId("6400ae417fa60685f6b68b42"), v: '1' },
  { _id: ObjectId("6400ae417fa60685f6b68b43"), v: 1 },
  { _id: ObjectId("6400ae417fa60685f6b68b44"), v: 1 },
  { _id: ObjectId("6400ae417fa60685f6b68b45"), v: 1 },
  { _id: ObjectId("6400ae417fa60685f6b68b46"), v: Long("1") },
  { _id: ObjectId("6400ae417fa60685f6b68b47"), v: 1 },
  { _id: ObjectId("6400ae587fa60685f6b68b48"), v: 1 },
  { _id: ObjectId("6400ae717fa60685f6b68b49"), v: 1 },
  { _id: ObjectId("6400aeeb7fa60685f6b68b4a"), v: 1 },
  { _id: ObjectId("6400af347fa60685f6b68b4b"), v: 1 }
]
```

I even added more examples (such as v:1.1), and afterwards used `{$inc{v:-0.1}}`, and it seems that every time when the value can be decimal, it converts to decimal. I did not occur any diffs between FerretDB and MongoDB, while running `$ne` on those cases.

Reply:
Thanks a lot for the detailed testing!